velox/functions/sparksql/aggregates/AverageAggregate.cpp

```diff
@@ -96,13 +373,16 @@ exec::AggregateRegistrationResult registerAverage(const std::string& name) {
                              .build());
   }
 
-  signatures.push_back(exec::AggregateFunctionSignatureBuilder()
-                           .integerVariable("a_precision")
-                           .integerVariable("a_scale")
-                           .argumentType("DECIMAL(a_precision, a_scale)")
-                           .intermediateType("varbinary")
-                           .returnType("DECIMAL(a_precision, a_scale)")
-                           .build());
+  signatures.push_back(
+      exec::AggregateFunctionSignatureBuilder()
+          .integerVariable("a_precision")
+          .integerVariable("a_scale")
+          .integerVariable("r_precision", "min(38, a_precision + 4)")
+          .integerVariable("r_scale", "min(38, a_scale + 4)")
+          .argumentType("DECIMAL(a_precision, a_scale)")
+          .intermediateType("ROW(DECIMAL(38 , a_scale), BIGINT)")
+          .returnType("DECIMAL(r_precision, r_scale)")
+          .build());
```

Comment:
Below extra registration is needed for partial merge stage of average. In this case, the intermediate type is `ROW(DECIMAL(a_precision, a_scale), BIGINT)` and not restricted precision 38.

```
signatures.push_back(
      exec::AggregateFunctionSignatureBuilder()
          .integerVariable("a_precision")
          .integerVariable("a_scale")
          .argumentType("DECIMAL(a_precision, a_scale)")
          .intermediateType("ROW(DECIMAL(a_precision, a_scale), BIGINT)")
          .returnType("DECIMAL(a_precision, a_scale)")
          .build());
```

Reply:
@rui-mo The best way is to define an i_precision and obtain it through certain calculations. However, currently velox does not support setting integerVariable for the intermediate result of agg, because FunctionSignature::validate will not pass the intermediateType of agg into validate, and an exception error will occur.
```cpp
exec::AggregateFunctionSignatureBuilder()
          .integerVariable("a_precision")
          .integerVariable("a_scale")
          .integerVariable("r_precision", "min(38, a_precision + 4)")
          .integerVariable("r_scale", "min(38, a_scale + 4)")
          .integerVariable("i_precision", "min(38, a_scale + 10")
          .argumentType("DECIMAL(a_precision, a_scale)")
          .intermediateType("ROW(DECIMAL(i_precision , a_scale), BIGINT)")
          .returnType("DECIMAL(r_precision, r_scale)")
          .build());
```
```
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: (4 vs. 5) Some integer variables are not used
Retriable: False
Expression: usedVariables.size() == variables.size()
Function: validate
File: /root/remote_dev/velox/velox/expression/FunctionSignature.cpp
Line: 215
```
@mbasmanova Should we make validate as a member function so that AggregateFunctionSignature can implement its own validate function?